kroxylicious-test-tools/src/main/java/io/kroxylicious/test/client/KafkaClient.java

```diff
@@ -78,9 +84,31 @@ private static DecodedRequestFrame<?> toApiRequest(Request request) {
      * @return a future that will be completed with the response from the kafka broker (translated to JsonNode)
      */
     public CompletableFuture<Response> get(Request request) {
-        DecodedRequestFrame<?> decodedRequestFrame = toApiRequest(request);
+        return getAll(request).thenApply(response -> {
+            if (response.size() != 1) {
+                throw new IllegalStateException("expected one response");
+            }
+            return response.get(0);
+        });
+    }
+
+    /**
+     * Bootstrap and connect to a Kafka broker on a given host and port. Send
+     * the requests to it and inform the client when we have received all responses.
+     * The channel is closed after we have received the message.
+     * @param requests request to send to kafka
+     * @return a future that will be completed with the response from the kafka broker (translated to JsonNode)
+     */
+    public CompletableFuture<List<Response>> getAll(Request... requests) {
+        List<DecodedRequestFrame<?>> decodedRequestFrames = Arrays.stream(requests).map(KafkaClient::toApiRequest).collect(toList());
         CorrelationManager correlationManager = new CorrelationManager();
-        KafkaClientHandler kafkaClientHandler = new KafkaClientHandler(decodedRequestFrame);
+        // if we are sending multiple requests then we need to behave like the kafka client, sending an initial request and
+        // awaiting response, this prevents blowing up the kroxylicious state machine when there is more data available to
```

Comment:
That sounds like a separate Kroxylicious bug to me.  AFAIK there nothing in the Kafka Spec that says a client has to wait the ApiVersions response.  A client might know, a priori, the version of the Cluster, so could just start pipelining other requests down the wire. 

Kroxylicious will no doubt be hooked up to all sorts of different clients and intermediaries, so we want to assume as little as we can about behaviour.

Reply:
True, I could imagine fixing it by queuing further requests while we are in these pre-connected states.  Will make an issue.

Reply:
https://github.com/kroxylicious/kroxylicious/issues/513